c18n: New {get,set}context APIs for libunwind

dpgao opened the pull request:
See also #2122 for a draft of this PR.

_rtld_unw_getcontext_epilogue is removed because when c18n is not enabled, the trusted stack field of the context is unused anyway, so there's no need to fill it.

The assembly stub in _rtld_unw_setcontext is removed because _rtld_unw_setcontext is now expected to be called before all registers are restored, so it does not need to restore any register anymore.

dstolfa commented:
This makes changes in setjmp/longjmp as well and the commit message should probably reflect that.

bsdjhb commented on lib/libc/aarch64/gen/_setjmp.S:
```diff
@@ -81,9 +81,7 @@ ENTRY(_longjmp)
-#if defined(__CHERI_PURE_CAPABILITY__) && defined(RTLD_SANDBOX)
-	mov	c2, c8
-#else
+#if !defined(__CHERI_PURE_CAPABILITY__) || !defined(RTLD_SANDBOX)
```
I find it easier to read as !(purecap && sandbox) rather than !purecap || !sandbox.

bsdjhb replied:
Also, is RTLD_SANDBOX ever true for !purecap?  That is, can we simplify some of the #ifdef's to '#ifdef RTLD_SANDBOX` or `#ifndef RTLD_SANDBOX`?

dpgao replied:
> Also, is RTLD_SANDBOX ever true for !purecap? That is, can we simplify some of the #ifdef's to `#ifdef RTLD_SANDBOX` or `#ifndef RTLD_SANDBOX`?

The purecap check is indeed redundant. I'll do a find-and-replace once the libsys is merged. (libsys renames RTLD_SANDBOX to CHERI_LIB_C18N.)

dpgao replied:
@bsdjhb I've removed the checks for `__CHERI_PURE_CAPABILITY__` in #2200.

bsdjhb reviewed (commented):
Certainly some text in the commit log explaining why setjmp/longjmp are changing would be good.

dstolfa reviewed (commented):
Builds & runs fine, similar to https://github.com/CTSRD-CHERI/llvm-project/pull/740 just a few nits.

dpgao commented on libexec/rtld-elf/rtld_c18n.h:
```diff
@@ -120,6 +120,48 @@ struct stk_table {
+	/*
+	 * Compartment ID of the caller
+	 */
+	stk_table_index caller;
```
The comment above is actually misleading. `stk_table_index` and compartment IDs are different types, but they can be converted to and from each other. I will update the comment.

dpgao commented on libexec/rtld-elf/rtld_c18n.h:
```diff
@@ -120,6 +120,48 @@ struct stk_table {
+	 */
+	void *osp;
+	/*
+	 * Address of the previous trusted frame
```
Ack

jrtc27 commented on lib/libc/aarch64/gen/_setjmp.S:
```diff
@@ -81,9 +81,7 @@ ENTRY(_longjmp)
-	mov	c2, c8
-#else
+#ifndef CHERI_LIB_C18N
 	mov	REGN(sp), REG(8)
```
What's wrong with doing this for c18n? I'd like to see as much of libc etc de-ifdef'ed as eventually I'd like to see lib/c18n killed off, just as libunwind doesn't need two copies.

jrtc27 commented on libexec/rtld-elf/rtld_c18n.h:
```diff
@@ -217,8 +259,8 @@ func_sig_legal(struct func_sig sig)
  */
-#define	c18n_return_address()						\
-	(C18N_ENABLED ? get_trusted_stk()->pc : __builtin_return_address(0))
+#define	c18n_return_address()	(C18N_ENABLED ?				\
```
There should probably be a non-c18n fallback implementation of this so you don't need to ifdef every use.

dpgao replied:
Will do in another PR

jrtc27 commented on libexec/rtld-elf/rtld_c18n.c:
```diff
@@ -936,13 +913,17 @@ unwind_stack(struct jmp_args ret, void *rcsp, struct trusted_frame *target)
 	 * Make the function re-entrant by blocking all signals.
 	 */
 	SIGFILLSET(nset);
 	sigprocmask(SIG_SETMASK, &nset, &oset);
```
Could expose functions from rtld_lock.c that use sigfastblock when available, saves having to do syscalls here for single-threaded applications. In theory one could also register a per-thread mask just for c18n to use, given rtld unregisters its locking one when a thread is created. But not for this PR.

dpgao commented on libexec/rtld-elf/rtld_c18n.c:
```diff
@@ -859,16 +858,19 @@ resolve_untrusted_stk_impl(stk_table_index index)
 
-static void *
-unwind_cursor(void)
+bool c18n_is_enabled(void);
```
Yes. Although I guess it would require renaming `struct trusted_frame` etc. to have a `dl_` prefix?

dpgao replied:
Hmmm, since a `struct trusted_frame *` only ever leaves RTLD as a seal capability, a `void *` should be sufficient.

dpgao commented:
@jrtc27 I've changed how setjmp/longjmp invokes the RTLD's unwinding APIs such that it is very simple to remove the ifdefs in a future PR (essentially we just need a noop ASM stub like in libunwind). I also added `dl_` prefixes to the APIs and exported them in `link_elf.h`.

dpgao commented:
The `dl_*` functions now use the normal ABI and CTSRD-CHERI/llvm-project#740 has been updated accordingly.

jrtc27 commented:
Can we keep the old interfaces implemented as compatibility shims so upgrades work when c18n is enabled?

dpgao commented:
> Can we keep the old interfaces implemented as compatibility shims so upgrades work when c18n is enabled?

It seems that preserving the old interfaces implies retaining much of the old implementation as well (including the assembly stubs), but that's fine.

jrtc27 commented on sys/sys/link_elf.h:
```diff
@@ -110,6 +110,12 @@ int _rtld_addr_phdr(const void *, struct dl_phdr_info *);
 int _rtld_get_stack_prot(void);
 int _rtld_is_dlopened(void *);
 
+struct dl_c18n_compart_state;
```
Including the header should give this a complete type (presumably from some machine/foo.h) rather than a forward declaration

dpgao replied:
Ack

jrtc27 commented on libexec/rtld-elf/rtld_c18n.c:
```diff
@@ -882,17 +883,23 @@ unwind_cursor(void)
 }
 
+/*
+ * XXX: These functions are kept here for compatibility with old libunwind.
```
libc and libunwind (ditto for the other instance below)

jrtc27 replied:
And maybe group all the compat interfaces together at the end?

dpgao replied:
It is possible to move all the function definitions to the end, but the declaration of `_rtld_unw_setcontext_impl` has to stay in front because of the function-pointer workaround employed in the previous release. `_rtld_unw_setcontext_impl` has to be wrapped into `_rtld_unw_setcontext_ptr` in `c18n_init2`.

jrtc27 commented on libexec/rtld-elf/rtld_c18n.c:
```diff
@@ -867,8 +868,8 @@ resolve_untrusted_stk_impl(stk_table_index index)
-static void *
-unwind_cursor(void)
+void
+dl_c18n_get_trusted_stk(void **buf)
```
Why does it need to take a `void **buf`? Just return the value and let the caller do a store if it wants to?

dpgao replied:
This function is only called from assembly and this signature makes it a bit easier to use from assembly.

jrtc27 replied:
In your current implementation it happens to only be called from assembly in places where the assembly wants to store the result into memory. That's not necessarily a justification for making the interface unusual like this.

jrtc27 commented on lib/libgcc_s/Makefile:
```diff
@@ -53,10 +53,4 @@ SRCS+=		s_logbl.c
-# LIBUNWIND_SANDBOX_OTYPES is only supported on aarch64 (Morello).
-.if ${MACHINE_ABI:Mpurecap} && ${MACHINE_CPUARCH} == "aarch64"
-SYMBOL_MAPS+=	${.CURDIR}/Symbol-c18n.map
-CFLAGS+=	-D_LIBUNWIND_CHERI_C18N_SUPPORT
```
Where's the -D_LIBUNWIND_HAS_CHERI_LIB_C18N?

jrtc27 commented on lib/libc/gen/dlfcn.c:
```diff
@@ -377,4 +377,28 @@ _rtld_is_dlopened(void *arg __unused)
+	return (NULL);
+}
+
+
```
```suggestion
```

jrtc27 commented on sys/sys/link_elf.h:
```diff
@@ -110,6 +110,14 @@ int _rtld_addr_phdr(const void *, struct dl_phdr_info *);
 
+#if defined(__CHERI_PURE_CAPABILITY__) && defined(__aarch64__)
+#include <machine/c18n.h>
+void *dl_c18n_get_trusted_stk(void);
```
Please just spell stack in full in the public API, you don't need to save two letters

bsdjhb commented:
I'm in the middle of another FreeBSD merge, but I can do an update of libunwind after that so long as the new libunwind will still work fine with the current `dev`

jrtc27 commented:
Yeah, it would be easiest if we decoupled rtld and libunwind changes. Because this PR keeps the old rtld interfaces (and they’ll stay for one release) we don’t need to import libunwind in this PR, we can just merge the new interfaces ready for a future import.

jrtc27 commented:
(But note that updating libunwind in dev now would break)

dpgao commented:
> Be very careful about git subrepo commits in a PR, the parent commit is recorded so rebasing breaks them for next time someone uses git subrepo.

I see. I've removed the subrepo pull commit.

jrtc27 reviewed (approved):
I think this is ok now, don't know if someone else wants to give it a once over before merging

dpgao commented:
Pinging @bsdjhb
Is this a good time to merge now?

bsdjhb commented:
I'm in the middle of a FreeBSD merge, but I've tagged it as ready to land and will land it after the next merge.
